Refactor core::stream, build a shared lib

FloopCZ opened the pull request:
~~This is a **heavy WIP**, don't use, don't look, don't complain.~~

Backward incompatible changes:
- `stream::batch` is renamed to `stream::rebatch`.
- There is no `HiPipe::Python` CMake target anymore, Python functionality is integrated into `HiPipe::Core`.
- `stream::buffer` and `stream::rebatch` cannot be followed by any more stream transformers and have to be the last operations in the pipeline. This is for performance, see the documentation for more details.
- After #17, `dataframe` is not a template anymore.

**Right before merging this one, merge #12 #15 #17.**

**DO NOT SQUASH**

A little compile time statistics of one of our larger projects.

Before (`dev` branch):
```
        Elapsed (wall clock) time (h:mm:ss or m:ss): 2:12.58
        Maximum resident set size: 8.57 GB
        Major (requiring I/O) page faults: 180
        File system inputs: 62016
        File system outputs: 238528
```

After (`fill-shared-lib` branch):
```
        Elapsed (wall clock) time (h:mm:ss or m:ss): 1:30.61
        Maximum resident set size: 6.05 GB
        Major (requiring I/O) page faults: 2
        File system inputs: 4248
        File system outputs: 163944
```

This was not the main purpose of this pull request, but the improvement in compilation time and memory consumption seems significant.

This PR changes the representation of the stream itself - previously, the type of the stream changed after each transformation, now it stays the same - `forward_stream_t` (which is actually an alias to `ranges::any_view<batch_t>`). I was kind of hoping this feature will allow a new syntax for very long pipelines that will push the compilation time even lower, but I have not experimented with that yet. Some other time hopefully.

FloopCZ committed:
Use c++17

FloopCZ committed:
Refactor column

FloopCZ committed:
Refactor stream::create

FloopCZ committed:
Generalize batch insert

FloopCZ committed:
Line breaks

petrbel commented on include/hipipe/core/stream/column.hpp:
```diff
@@ -54,44 +106,90 @@ class column_base {
+    Column& at()
+    {
+        throw_check_extraction_type<Column>();
+        return *columns_.at(std::type_index{typeid(Column)});
```
why not `columns_->at` ?

FloopCZ replied:
That would be a different thing. `columns_` is not a pointer, it is a `map`. I want to dereference the pointer retrieved by `at()`. Operator `*` has a lower priority than `.`.

petrbel commented on include/hipipe/core/stream/column.hpp:
```diff
@@ -54,44 +106,90 @@ class column_base {
+    // column check //
+
+    template<typename Column>
+    bool contains() const
```
so 2020

FloopCZ committed:
Allow noncopyable columns

FloopCZ committed:
Refactor partial_transform

FloopCZ committed:
Add partial_transform test

FloopCZ committed:
Comment out failing code

FloopCZ committed:
Refactor stream::transform [needs tests]

FloopCZ committed:
Do not pass NOuts in transform

FloopCZ committed:
Refactor transform1.cpp tests

FloopCZ committed:
Add batch constructors for better error messages

FloopCZ committed:
Refactor transform2.hpp

FloopCZ committed:
Refactor transform3.hpp

FloopCZ committed:
Refactor transform4.hpp

FloopCZ committed:
Refactor transform5.hpp

FloopCZ committed:
Minor update to generate and random_fill

FloopCZ committed:
Refactot random_fill

FloopCZ committed:
Minor improvements to column

FloopCZ committed:
Clear common includes

FloopCZ committed:
Add some static assertions

FloopCZ committed:
Refactor generate and random fill

FloopCZ committed:
Refactor stream::pad

FloopCZ committed:
Refactor unpack

FloopCZ committed:
Refactor for_each

FloopCZ committed:
Refactor stream::drop

FloopCZ committed:
Refactor stream::buffer

FloopCZ committed:
Buffer read provides rvalue references

FloopCZ committed:
Make buffer single pass

FloopCZ committed:
Rework stream::batch

FloopCZ committed:
Allow InputRanges

FloopCZ committed:
Refactor stream::filter

FloopCZ committed:
Revert test_ranges_equal function

FloopCZ committed:
Add some forgotten rebranding stuff

FloopCZ committed:
Refactor CMakeLists

- from now on, hipipe_core is not a header-only library

FloopCZ committed:
std::apply is not experimental anymore

FloopCZ committed:
Add is_specialization helper function

FloopCZ committed:
Add column and batch to_python converters

FloopCZ committed:
GCC cannot handle maybe_tuple return type, workaround

FloopCZ committed:
Refactor python stream converter

FloopCZ committed:
Inline a function that should be inline.

FloopCZ committed:
Various little things I missed

FloopCZ committed:
Update range-v3 submodule

FloopCZ committed:
Update .gitignore

FloopCZ committed:
Update range-v3 submodule

FloopCZ committed:
Make stream an input range

FloopCZ committed:
s/utility-vector/utility-ndim/

FloopCZ committed:
Buffer requires forward range

FloopCZ committed:
Remove some useless inlines

FloopCZ committed:
Simplify partial_transform_fn

FloopCZ committed:
Add more static assertion + minor optimizations

FloopCZ committed:
Handle recursive ranges gracefully in ndims<>

FloopCZ committed:
Improve docstring for batch and buffer

FloopCZ committed:
Separate stream_t to forward and input stream_t

FloopCZ committed:
Erase transform function type

FloopCZ committed:
Add function type erasure for stream::filter

FloopCZ committed:
Add function type erasure for stream::transform

FloopCZ committed:
Properly constify std::function params for stream::filter

FloopCZ committed:
Add static assert and type erasure to for_each

FloopCZ committed:
Add type erasure to generate and random_fill

FloopCZ committed:
Replace short ranges pipelines by function-like syntax

FloopCZ committed:
Update .gitignore

FloopCZ committed:
Move python folder hierarchy to core

FloopCZ committed:
Remove tuple_remove and dependency on boost::hana

FloopCZ committed:
Use fully qualified names in filter tests

FloopCZ committed:
Rename stream::batch to stream::rebatch

FloopCZ committed:
GIve HIPIPE_ prefix to all CMake variables

FloopCZ committed:
Minor updates to .circleci/config.yml

FloopCZ committed:
Move python tests under core tests

FloopCZ committed:
Separate common test header for stream and the rest of the core

FloopCZ committed:
Clean up column, batch and stream objects

FloopCZ committed:
Require ForwardRange for stream::create

FloopCZ committed:
Remove unused ref_wrap_t

FloopCZ committed:
Remove some constexpr-s

FloopCZ committed:
Rename column.value() to column.data()

FloopCZ committed:
Delete unused insert function in batch

FloopCZ committed:
Rename batch::insert to batch::insert_or_assign

FloopCZ committed:
Simplify column constructor

FloopCZ committed:
Rename column::batch_type to column::data_type to avoid confusion

FloopCZ committed:
Remove TODO

FloopCZ committed:
Cleanup stream::create tests

FloopCZ committed:
Add stream, batch and column diagram

FloopCZ committed:
Reformulate the documentation

FloopCZ committed:
Add .gitattributes

FloopCZ committed:
Fix flags in installation guide

FloopCZ committed:
Fix multiple docstrings

FloopCZ committed:
Add stubs for batch and column tests

Commit:
WIP column tests

Commit:
batch tests & requested changes

Commit:
test fix

Commit:
remove unused functions

FloopCZ committed:
Merge pull request #13 from iterait/delete-unused-functions

remove unused functions

FloopCZ committed:
Check before column push_back

FloopCZ committed:
Merge branch 'column-batch-tests' of github.com:iterait/hipipe into column-batch-tests

FloopCZ committed:
Remove some unnecessary stuff

FloopCZ committed:
Avoid multiple RTTI checks

FloopCZ committed:
Change the tested exception type

Commit:
requested changes

Commit:
Merge branch 'column-batch-tests' of github.com:iterait/hipipe into column-batch-tests

FloopCZ committed:
Merge branch 'dev' into runtime

FloopCZ committed:
Merge branch 'dev' into runtime

FloopCZ committed:
Remove unnecessary constexpr

FloopCZ committed:
Remove shared lib placeholder

FloopCZ committed:
Move base64 to shared lib

FloopCZ committed:
Move csv utils to shared library

FloopCZ committed:
Put python stream converter to shared library

FloopCZ committed:
Move utility/filesystem to shared library

FloopCZ committed:
Dataframe is not a template anymore

This needed some code reshuffling to make sure that
functions are defined before use.

FloopCZ committed:
Remove documentation from .cpp file

Commit:
Merge pull request #12 from iterait/column-batch-tests

Column batch tests

Commit:
Merge pull request #17 from iterait/fill-shared-lib

Move many functions to the shared library

blazekadam reviewed (approved):
God bless thy letter!
